odo/into.py

```diff
@@ -1,23 +1,40 @@
 from __future__ import absolute_import, division, print_function
 
-from toolz import merge
+from toolz import merge, curry
 from multipledispatch import Dispatcher
 from .convert import convert
 from .append import append
 from .resource import resource
 from .utils import ignoring
+
+import datashape
 from datashape import discover
 from datashape.dispatch import namespace
 from datashape.predicates import isdimension
+
 from .compatibility import unicode
 
 
-if 'into' not in namespace:
-    namespace['into'] = Dispatcher('into')
-into = namespace['into']
+__all__ = 'into',
+
+
+if '_into' not in namespace:
+    namespace['_into'] = Dispatcher('_into')
+
+_into = namespace['_into']
 
 
-@into.register(type, object)
+@curry
+def into(a, b, dshape=None, **kwargs):
+    if isinstance(dshape, (str, unicode)):
+        dshape = datashape.dshape(dshape)
+    assert isinstance(dshape, datashape.DataShape) or dshape is None, \
```

Comment:
could this be a typeerror? imho typeerror better conveys that the user passed the incorrect type to the function